### PR TITLE
Add proper deregistering of events & removal of elements for amp-youtube component

### LINKED ISSUE #7720: amp-youtube should free resources upon unlayout
### First Timers Only
We know figuring out the process for contributing to an open source project can be intimidating, so we created this issue as a way for you to learn the ropes.  (If you feel comfortable contributing to open source projects, please leave this issue for someone else.)

This particular starter issue is designed to familiarize you with the AMP Code Contribution process and to show-case one of the basic life-cycle events of an AMP component.

### Background
AMP runtime manages the life-cycle of components ensuring they are built, layed-out and garbage collected at the right time to provide stellar performance.

Each element implements their functionality for each life-cycle hook by overriding several `callback` methods such as:
 - `buildCallback`: Called when it is time to initialize a component. 
 - `layoutCallback`: Called when a component is about to come to view. This is where a component does most of its heavy work.
 - `unlayoutCallback`: Called when it is time to tear-down a component to free expensive resources.

A complete list of life-cycle hooks is documented in [Building an AMP Extension document](https://docs.google.com/document/d/19o7eDta6oqPGF4RQ17LvZ9CHVQN53whN-mCIeIMM8Qk).

### Motivation
Currently `amp-youtube` component does not override `unlayoutCallback`. This means resources (such as event listeners and the actual fairly-heavy YouTube iframe) are not freed back to the system.

### The bug
 `amp-youtube` should implement `unlayoutCallback` which:
- [ ] unlistens the `message` listener it registers on window
- [ ] removes the iframe element

`amp-instagram` component has a good implementation for its `unlayoutCallback` and can be used as a reference implementation for this issue.

### Step by step
- [ ] Claim this issue by adding a comment below.  Please only claim this bug if you plan on starting work in the next day or so.
- [ ] If you aren't too familiar with Git/GitHub, see the [Getting Started End-to-End Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md) for [an intro to Git & GitHub,](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#intro-to-git-and-github) and [how to get a copy of the code](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#get-a-copy-of-the-amphtml-code).  You can also refer to the [Quick Start Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) for the necessary setup steps with less explanation than the End-to-End guide.
- [ ] Follow the instructions for [building AMP](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#building-amp-and-starting-a-local-server).
- [ ] [Create a Git branch](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#create-a-git-branch) for making your changes.
- [ ] [Commit your changes](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#edit-files-and-commit-them) frequently.
- [ ] [Push your changes to GitHub](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#push-your-changes-to-your-github-fork).
- [ ] [Create a Pull Request](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review).  Mention `closes Issue #7720` in the description.
- [ ] Assign @aghassemi as the reviewer
- [ ] [Respond to your reviewer's comments](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#respond-to-pull-request-comments) (if any).

Once approved, your changes will be merged.  **⚡⚡⚡Congrats on making your first contribution to the AMP Project!⚡⚡⚡**  You'll be able to see it [live across the web soon](https://github.com/ampproject/amphtml/blob/master/contributing/release-schedule.md)!

Thanks, and we hope to see more contributions from you soon.

### Questions?

If you have questions ask in this issue or on your Pull Request (mentioning @aghassemi) or see the [How to get help](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#how-to-get-help) section of the Getting Started guide.

### PR DESCRIPTION
**Pull request by ldthorne**
Closes #7720.

Work in progress.

### CODE REVIEW

**Comment by ldthorne** [comment]
I signed it!

File: extensions/amp-youtube/0.1/amp-youtube.js [hunk 1]
```diff
@@ -190,17 +195,30 @@ class AmpYoutube extends AMP.BaseElement {
...
+      this.handleYoutubeMessages_.bind(this)
+    );
 
     return this.loadPromise(iframe)
         .then(() => this.listenToFrame_())
         .then(() => this.playerReadyPromise_);
   }
 
   /** @override */
+  unlayoutCallback() {
```
Comment (jridgewell) [inline extensions/amp-youtube/0.1/amp-youtube.js]: We also need to reset `#playerState`, `#playerReadyPromise`, and `#playerReadyResolver`.

File: extensions/amp-youtube/0.1/test/test-amp-youtube.js [hunk 2]
```diff
@@ -291,6 +291,17 @@ describe('amp-youtube', function() {
     });
   });
 
+  it('should remove iframe after unlayoutCallback', () => {
+    return getYt({'data-videoid': 'mGENRKrdoGY'}).then(yt => {
+      const placeholder = yt.querySelector('[placeholder]');
+      const obj = yt.implementation_;
+      obj.unlayoutCallback();
+      expect(yt.querySelector('iframe')).to.be.null;
```
Comment (aghassemi) [inline extensions/amp-youtube/0.1/test/test-amp-youtube.js]: let's also make sure `unlistenMessage_` is called. You can do that by installing a `spy` (e.g. `const unlistenSpy = sandbox.spy(obj, 'unlistenMessage_')`) on it before calling  `unlayoutCallback` and then checking that spy has been called via `expect(unlistenSpy).to.have.been.called;`

File: extensions/amp-youtube/0.1/test/test-amp-youtube.js [hunk 3]
```diff
@@ -291,6 +291,17 @@ describe('amp-youtube', function() {
...
+      expect(obj.iframe_).to.be.null;
```
Comment (aghassemi) [inline extensions/amp-youtube/0.1/test/test-amp-youtube.js]: let's also assert that `#playerState, #playerReadyPromise, and #playerReadyResolver` that @jridgewell suggested to add are reset here.

**Comment by ldthorne** [comment]
Hey @aghassemi @jridgewell, sorry for the delay in getting finishing this up -- I think it's ready for another look.

File: extensions/amp-youtube/0.1/amp-youtube.js [hunk 4]
```diff
@@ -186,15 +191,41 @@ class AmpYoutube extends AMP.BaseElement {
 
     this.iframe_ = iframe;
 
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleYoutubeMessages_.bind(this)
+    );
     this.win.addEventListener(
```
Comment (jridgewell) [inline extensions/amp-youtube/0.1/amp-youtube.js]: This is no longer necessary with the `listen` above.

File: extensions/amp-youtube/0.1/amp-youtube.js [hunk 5]
```diff
@@ -186,15 +191,41 @@ class AmpYoutube extends AMP.BaseElement {
...
   /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+    if (this.playerState_ !== PlayerStates.PAUSED) {
```
Comment (jridgewell) [inline extensions/amp-youtube/0.1/amp-youtube.js]: We can skip the if and just set player state to `PAUSED`.

File: extensions/amp-youtube/0.1/amp-youtube.js [hunk 6]
```diff
@@ -186,15 +191,41 @@ class AmpYoutube extends AMP.BaseElement {
...
+      this.playerState_ = PlayerStates.PAUSED;
+    }
+    if (this.playerReadyPromise_) {
+      this.playerReadyPromise_ = null;
```
Comment (jridgewell) [inline extensions/amp-youtube/0.1/amp-youtube.js]: This needs to be reset to a pending promise.

File: extensions/amp-youtube/0.1/amp-youtube.js [hunk 7]
```diff
@@ -186,15 +191,41 @@ class AmpYoutube extends AMP.BaseElement {
...
+    }
+    if (this.playerReadyResolver_) {
+      this.playerReadyResolver_ = null;
```
Comment (jridgewell) [inline extensions/amp-youtube/0.1/amp-youtube.js]: We can combine this into the previous if statement. It also needs to be set to the pending promise's resolve function.

File: extensions/amp-youtube/0.1/amp-youtube.js (hunk 6 above)
Comment (ldthorne) [inline extensions/amp-youtube/0.1/amp-youtube.js]: @jridgewell both with regard to the above comment and the next, are you thinking something along the lines of what's done on line ~122:
```js
this.playerReadyPromise_ = new Promise(resolve => {
  this.playerReadyResolver_ = resolve;
});
```

Comment (aghassemi) [inline extensions/amp-youtube/0.1/amp-youtube.js]: @ldthorne, Justin is our of office, to answer your question: Yes. both `if`s can be replaced by the code you have above.

**Comment by ldthorne** [comment]
Hey @aghassemi, my build is failing due to the following error:
`A pull request may not contain a mix of flag-config and non-flag-config files. Please make your changes in separate pull requests. First offending file: .travis.yml`. My PR doesn't touch `.travis.yml`. Any ideas what might be going on?

**Comment by aghassemi** [comment]
@ldthorne Please rebase. A change went in a few weeks ago that required all open PRs to rebase to the HEAD of master.

**Comment by ldthorne** [comment]
Hey @aghassemi, thanks for the heads up. I've just rebased and fixed the specs. In the future, is there a place something like this is announced?

**Comment by aghassemi** [comment]
@ldthorne Your PR was just merged. Thanks for contributing! 

We did announcements on individual PRs on case-by-case basis since these changes should be fairly rare going forward.

**Pull request merged**